crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -232,49 +300,82 @@ impl SignatureMPCManager {
         Ok(())
     }
 
-    pub fn handle_mpc_message(
+    pub async fn handle_end_of_delivery(&mut self) -> PeraResult {
+        let txs_to_send: Vec<ConsensusTransaction> = self
+            .mpc_instances
+            .iter_mut()
+            .filter(|(_, instance)| {
+                instance.status == MPCSessionStatus::Active
+                    && instance.pending_messages.len() >= self.mpc_threshold_number_of_parties
+            })
+            .collect::<Vec<_>>()
+            .par_iter_mut()
+            .filter_map(|(_, ref mut instance)| instance.advance())
```

Comment:
Two questions about this line:
1. Does it sequentially advance each Instance one-by-one? Why not spin off a new thread for each Instance so that the
   current thread can continue to the next job?
2. The `filter_map` documentation says that it returns "a new iterator with only the items from `Some` results", are we
   not forgetting about any errors by dropping the Instances for which `advance()` returned `None`? Or are these caught
   later or differently?

Reply:
They run in parallel, I'll add a todo here on error handling & punishing malicious actors.